{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "86cffca18316bf6a",
   "metadata": {},
   "source": [
    "# Import configuration and setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "initial_id",
   "metadata": {},
   "outputs": [],
   "source": [
    "from medulloblastoma.config import PROJ_ROOT, RAW_DATA_DIR, INTERIM_DATA_DIR, PROCESSED_DATA_DIR, EXTERNAL_DATA_DIR, MODELS_DIR, REPORTS_DIR, FIGURES_DIR\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "os.chdir(PROJ_ROOT)\n",
    "from medulloblastoma.dataset import download_data, prepare_data\n",
    "from medulloblastoma.features import main as preprocess_pipeline\n",
    "from medulloblastoma.features import load_data\n",
    "# Execute R script for GSE85217 data download\n",
    "# ! Rscript {os.path.join(PROJ_ROOT, 'medulloblastoma','get_data.R')}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0821170af61b889",
   "metadata": {},
   "source": [
    "# Prepare Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d75c53e9da6194f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download microarray gene expression data\n",
    "download_data(save_path=RAW_DATA_DIR,remove_gz=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd204dcf95f01e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Structure data so that it is easier to handle\n",
    "prepare_data(\n",
    "    expression_file=os.path.join(RAW_DATA_DIR,'GSE85217_M_exp_763_MB_SubtypeStudy_TaylorLab.txt'),\n",
    "    metadata_path = os.path.join(RAW_DATA_DIR,'GSE85217_metadata.csv'),\n",
    "    save_path=RAW_DATA_DIR\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca5eb06740b4a455",
   "metadata": {},
   "source": [
    "# Preprocessing\n",
    "\n",
    "Selecting genes that are lowly expressed, lowly variant, and outlier genes. We also check that there are no missing data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bef6213f999ccb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocess_pipeline(\n",
    "    data_path=os.path.join(RAW_DATA_DIR,'cavalli.csv'),\n",
    "    metadata_path=os.path.join(RAW_DATA_DIR,'cavalli_subgroups.csv'),\n",
    "    save_path=PROCESSED_DATA_DIR,\n",
    "    per=0.2,\n",
    "    cutoff=0.1,\n",
    "    alpha=0.05\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aaf2057-6dd4-41d6-a30f-a79183c48235",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load preprocessed gene expression data\n",
    "data=pd.read_csv(os.path.join(PROCESSED_DATA_DIR,'cavalli_maha.csv'),index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24495fa6-15f1-421e-82f7-05e9a290c26a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata\n",
    "metadata=pd.read_csv(os.path.join(RAW_DATA_DIR,'cavalli_subgroups.csv'),index_col=0).squeeze()\n",
    "# Adapt names of groups of interests so they are shorter\n",
    "metadata=metadata.map({'Group3':'G3','Group4':'G4'})\n",
    "metadata.name = 'Subgroup'\n",
    "# Select groups of interest in metedata\n",
    "metadata_g3g4=metadata[metadata.isin(['G3','G4'])]\n",
    "print(metadata.shape,metadata_g3g4.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b68567-fa8a-4b14-bb20-099e8d7c9a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select groups of interest in gene expression data\n",
    "data_g3g4 = data.loc[metadata_g3g4.index]\n",
    "print(data_g3g4.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed07d3c3-19a9-4f8c-812a-c11316b84957",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0deb88a7-2f0b-426f-ae16-0bd8d2e0cb98",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_g3g4.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30da1a098957c480",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Load final datasets\n",
    "# data,metadata=load_data(\n",
    "#     data_path=os.path.join(PROCESSED_DATA_DIR,'cavalli_maha.csv'),\n",
    "#     metadata_path=os.path.join(PROCESSED_DATA_DIR,'g3g4_maha.csv')\n",
    "# )\n",
    "# print(data.shape,metadata.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3712080bddfd14fe",
   "metadata": {},
   "source": [
    "# UMAP Visualization of Preprocessed Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cc2e8db6ae70e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "from medulloblastoma.plots import plot_umap_binary\n",
    "\n",
    "# Discrete color mapping for G3/G4 subtypes\n",
    "dict_medulloblastoma = {\n",
    "    'G3': 'red',  # Red for G3\n",
    "    'G4': 'blue'   # Blue for G4\n",
    "}\n",
    "\n",
    "# Generate UMAP with discrete subtype coloring\n",
    "plot_umap_binary(\n",
    "    data=data_g3g4,\n",
    "    clinical=metadata_g3g4,\n",
    "    colors_dict=dict_medulloblastoma,\n",
    "    n_components=2,\n",
    "    save_fig=False,\n",
    "    save_as=\"initial_medulloblastoma_umap\",\n",
    "    seed=2023,\n",
    "    title=\"Medulloblastoma G3/G4 Gene Expression UMAP\",\n",
    "    marker_size=20\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "449de4cac0b4df86",
   "metadata": {},
   "source": [
    "# Model Training and Reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b369f9c1dadc9cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BLANK SECTION FOR MODEL TRAINING\n",
    "# This section will be implemented during the hackathon\n",
    "# Plan for:\n",
    "# - CVAE architecture definition\n",
    "# - Training loop with G3/G4 labels\n",
    "# - Hyperparameter optimization\n",
    "# - Model evaluation and validation\n",
    "# Clue: for architecture and hyperparameter optimization, you can use ax (see https://ax.dev/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97ae969b805f63af",
   "metadata": {},
   "source": [
    "# Final UMAP with Continuous Scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8064e60618e9030b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from medulloblastoma.plots import plot_umap_spectrum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "736c6f91-a55a-4705-b334-7a3da7eb7f6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For testing purposes, we assign a random score between 0 and 1 to each patient\n",
    "score=pd.Series(np.random.rand(470),name='score',index=metadata_g3g4.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4897aaf7b024ea1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_umap_spectrum(\n",
    "    data=data_g3g4,\n",
    "    clinical=score,  # Continuous scores instead of discrete labels\n",
    "    colormap='RdBu',\n",
    "    n_components=2,\n",
    "    save_fig=False,\n",
    "    save_as=\"final_scored_medulloblastoma_umap\",\n",
    "    seed=2023,\n",
    "    title=\"Medulloblastoma G3/G4 Scores\",\n",
    "    marker_size=20\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6b23b9c-5b9d-4051-b20c-d2243353820e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ML",
   "language": "python",
   "name": "ml"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
